Fix issue 13632: Enhancement to std.string.strip

aravindavk opened the pull request:
Added second argument similar to Python `str.strip`

Second argument accepts list of chars to strip and strips only
those chars.

Examples:

    "xyzhello".stripLeft("xyz") == ""hello"
    "helloxy ".stripRight("xy ") == "hello"
    "xhellox".strip("x") == "hello"

Signed-off-by: Aravinda VK <mail@aravindavk.in>

wilzbach commented on std/string.d:
```diff
@@ -2840,15 +2841,17 @@ if (isForwardRange!Range && isSomeChar!(ElementEncodingType!Range) &&
         if (std.ascii.isASCII(c))
         {
-            if (!std.ascii.isWhite(c))
+            if ((chars == "" && !std.ascii.isWhite(c)) ||
```
`is null` is the more idiomatic D way to check for empty strings.

aravindavk replied:
Thanks for review, I will update the patch to use null and refresh the patch.

jmdavis replied:
`is null` is most definitely _not_ the way to check for empty strings. That's how you check that a string is actually `null`. A `null` string is empty, but an empty string isn't necessarily null. The idiomatic way to check for empty is the `empty` function.

quickfur replied:
Yes, please don't use null to check for empty string, as it will introduce subtle bugs where empty strings with a non-null `.ptr` will not get recognized as empty.  Either use `.empty` or do an explicit length check `chars.length == 0`.

aravindavk commented on std/string.d:
```diff
@@ -2827,7 +2828,7 @@ if (isConvertibleToString!Range)
         Generic stripping on ranges: $(REF _stripLeft, std, algorithm, mutation)
   +/
-auto stripLeft(Range)(Range input)
+auto stripLeft(Range)(Range input, string chars = "")
```
Makes sense.  I will work on changing this as overloaded function as you suggested. Thanks for the review

burner commented:
I'm pretty sure that is not @jmdavis had in mind when he said overload.
This api is way to easy to use wrong.
Please, split it into two functions.

jmdavis commented:
Yeah, this is way too error-prone. The fact that you're providing a second function argument should be what determines whether the normal version of strip is used or not, not whether you passed a template argument _and_ passed a second function argument. The simplest way to deal with this is to just create two separate functions:

    auto stripLeft(Range)(Range input)
    if(...)
    {...}

    auto stripLeft(Range)(Range input, string chars)
    if(...)
    {...}

That way, the function signatures are clear, and the efficiency of the existing `stripLeft` is not affected by the new one.

aravindavk commented:
> Yeah, this is way too error-prone. The fact that you're providing a second function argument should be what determines whether the normal version of strip is used or not, not whether you passed a template argument and passed a second function argument. The simplest way to deal with this is to just create two separate functions:

Using second argument as compile time switch didn't worked since value should be known during compile time.

Without duplicating the function content I couldn't achieve two overloaded functions. Template argument is used in the intermediate function, both `strip(Range)` and `strip(Range, chars)` calls that function.

    auto stripLeft(Range, bool useCustomChars)(Range input, string chars)

    auto stripLeft(Range)(Range input)
    if(...)
    {
        return stripLeft(Range, false)(input)
    }

    auto stripLeft(Range)(Range input, string chars)
    if(...)
    {
         return stripLeft(Range, true)(input, chars)
    }

First function is internal function, I have to explore how it can be made private. Other two functions are exposed which doesn't use template parameter.

Thanks.

jmdavis commented:
> Without duplicating the function content I couldn't achieve two overloaded functions.

Just create two completely separate functions. It's simpler. e.g.

    auto stripLeft(R)(R range)
    if(...)
    {
        ...
    }

    auto stripLeft(R)(R range, string chars)
    if(...)
    {
        ...
    }

wilzbach commented:
> Without duplicating the function content I couldn't achieve two overloaded functions.

To avoid duplication, you could use a templated `stripLeftImpl`:

```d
auto stripLeft(R)(R range)
if(...)
{
    stripLeftImpl!(Yes.useChars)(range, chars);
}

auto stripLeft(R)(R range, string chars)
if(...)
{
    stripLeftImpl!(No.useChars)(range, chars);
}

private stripLeftImpl(Flag!"UseChars" useChars, R, Args...)(R range, Args args)
{
   ... code ...
   static if (useChars)
   {
       ...
   }
}
```

jmdavis commented:
> To avoid duplication, you could use a templated stripLeftImpl:

If you go that route, the flag is pointless. The number of arguments gives you the information you need. There is zero need to pass any kind of template argument indicating which version of the function to use.

jmdavis commented:
Looking over these again, I definitely vote for separate functions. Each overload should be doing fundamentally different things, and honestly, the fact that you've just swapped out `isWhite` for a call to `indexOf` implies that you don't really understand what the implementations of `stripLeft` and `stripRight` are doing.

With the new overload, for `stripLeft`, it's going to need to decode each character and compare it against the characters in `chars`. There's no point in the checks for `isASCII` and all of that, not without getting into checking that _all_ of the characters in `chars` are ASCII and then having different branches depending on that information.

And with `stripRight`, that big long set of if-elses has to do with looking for Unicode whitespace efficiently and has _nothing_ to do with `chars`

Just create completely separate overloads for these versions of `stripLeft` and `stripRight`, and don't just copy the existing ones and tweak them. How would you write these two functions if you weren't concerned about the existing overloads? Write their implementations as if the current ones didn't exist, and we can see what that would look like. The way that you've tried to tweak the existing ones really doesn't make sense. All of the fancy stuff in there is to make dealing with checking for whitespace more efficient and doesn't belong in overloads that aren't stripping whitespace.

aravindavk commented:
Thanks @jmdavis for detailed review, will work on the suggested changes and refresh the patch by today.

aravindavk commented:
Updated the code as suggested. Please review.

aravindavk commented:
Build is failing in CI, looks like unable to find `libdparse`,

    + make -f posix.mak -j2 publictests DUB=dub BUILD=debug
    make -C ../druntime -f posix.mak MODEL=64 DMD=../dmd/generated/linux/debug/64/dmd OS=linux BUILD=debug
    mkdir -p generated/linux/debug/64/publictests || exists generated/linux/debug/64/publictests
    make[1]: Entering directory '/home/circleci/druntime'
    touch generated/linux/debug/64/publictests/.directory
    make[1]: Nothing to be done for 'target'.
    make[1]: Leaving directory '/home/circleci/druntime'
    DFLAGS="-conf= -I../druntime/import  -w -de -dip25 -m64 -fPIC -g -debug generated/linux/debug/64/libphobos2.a -defaultlib= -debuglib= -L-ldl" dub build --force --compiler=${PWD}/../dmd/generated/linux/debug/64/dmd --single ../tools/tests_extractor.d
    Root package tests_extractor references unknown package libdparse
    posix.mak:610: recipe for target 'generated/linux/debug/64/tests_extractor' failed
    make: *** [generated/linux/debug/64/tests_extractor] Error 2
    rm generated/linux/debug/64/publictests/.directory
    Exited with code 2

quickfur commented on std/string.d:
```diff
@@ -2870,10 +2900,24 @@ if (isForwardRange!Range && isSomeChar!(ElementEncodingType!Range) &&
     assert(stripLeft([paraSep] ~ "hello world" ~ paraSep) ==
            "hello world" ~ [paraSep]);
 
+    assert(stripLeft("     hello world     ", " ") ==
```
Please put the new unittests in a separate unittest block, and also, please move the new `stripLeft` overload *after* the original unittest block. This is so that the generated documentation will make sense.

In other words, what you currently have is this:
````
/** ... */
auto stripLeft(...) ... // original stripLeft()

/** ... */
auto stripLeft(...) // new stripLeft

///
unittest
{
   ... // original test cases
   ... // new test cases
}
````

What it *should* be is this:

````
/** ... */
auto stripLeft(...) ... // original stripLeft

///
unittest { ... } // original test cases

/** ... */
auto stripLeft(...) ... // new stripLeft

///
unittest { ... } // new test cases
````

quickfur commented on std/string.d:
```diff
@@ -2882,6 +2926,12 @@ if (isConvertibleToString!Range)
     return stripLeft!(StringTypeOf!Range)(str);
 }
 
+auto stripLeft(Range)(auto ref Range str, string chars)
```
These two overloads also need to have the right order.  Here's what the order should be:
````
/** ... */
auto stripLeft(...) ... // original stripLeft

///
unittest { ... } // original test cases

/// ditto   <--- add this comment
auto stripLeft(Range)(auto ref Range str)
if (isConvertibleToString!Range) ...

///
unittest { ... } // more original test cases

/** ... */
auto stripLeft(...) ... // new stripLeft

///
unittest { ... } // new test cases

/// ditto   <--- add this comment
auto stripLeft(Range)(auto ref Range str, string chars)
if (isConvertibleToString!Range) ...

///
unittest { ... } // more new test cases (if there are any specific to the 2nd overload)
````

quickfur commented on std/string.d:
```diff
@@ -3010,6 +3107,13 @@ unittest
+    assert(stripRight("     hello world     ", " ") ==
+           "     hello world");
+    assert(stripRight("     hello worldxy     ", "xy ") ==
+           "     hello world");
```
Same goes for `stripRight`.  The original overload + unittests should be left as-is in their original order, and the new `stripRight` should be added *after* the original unittest, and the new tests should follow, in their own unittest block.

quickfur commented on std/string.d:
```diff
@@ -3092,6 +3236,12 @@ if (isConvertibleToString!Range)
+{
+    return strip!(StringTypeOf!Range)(str, chars);
+}
+
```
Same goes for these overloads.  Keep the original single-argument overloads grouped together with their respective unittest blocks, followed by the new 2-argument overloads and the new unittests.

aravindavk commented:
@quickfur thanks for the review, I will make the suggested changes and refresh the patch.

aravindavk commented:
@quickfur updated the patch as suggested. Please review

aravindavk commented:
Hi Reviewers, Let me know if any more changes required to the patch. Thanks.

quickfur commented on std/string.d:
```diff
@@ -3047,6 +3109,91 @@ if (isConvertibleToString!Range)
+  +/
+auto stripRight(Range)(Range str, string chars)
+if (isSomeString!Range ||
+    isRandomAccessRange!Range && hasLength!Range && hasSlicing!Range &&
```
Technically, we don't need to require a random access range here.  It is enough to require a bidirectional range and use `popBack` to strip from the end, in a similar way to `stripLeft`. In that case, we also don't need to require `hasLength` and `hasSlicing`, though of course in the implementation we can take advantage of those attributes where available.

quickfur commented on std/string.d:
```diff
@@ -3144,6 +3291,104 @@ if (isConvertibleToString!Range)
+  +/
+auto strip(Range)(Range str, string chars)
+if (isSomeString!Range ||
+    isRandomAccessRange!Range && hasLength!Range && hasSlicing!Range &&
```
ditto, we can relax the constraints here to just `isBidirectionalRange!Range`, and that should be sufficient to implement this function.

aravindavk commented:
@quickfur updated the code to use `isBidirectionalRange`, please review.

wilzbach commented on std/string.d:
```diff
@@ -2887,6 +2887,68 @@ if (isConvertibleToString!Range)
+    Returns: $(D input) stripped of extra characters.
+
+    Postconditions: $(D input) and the returned value
+    will share the same tail (see $(REF sameTail, std,array)).
```
Also you should use 'ditto' here and add the parameters to Ddoc header of the stripRight above 
This will allow Ddoc to group both functions.

aravindavk replied:
> Not sure whether this is really necessary. We typically just state that it doesn't allocate.

I followed this similar to stringLeft above.

Is there a way to generate the documentation locally to verify before sending patch? (Sorry if it is documented elsewhere)

aravindavk replied:
got it, testing documentation using `dmd -D std/string.d`, will refresh the patch with doc fixes. Thanks

wilzbach replied:
You can also preview the documentation at DAutoTest here or of course clone dlang.org and build it yourself

aravindavk commented:
@wilzbach tested doc generation locally(`dmd -D std/string.d` and `firefox string.html`) and modified doc text as you suggested. Please review.

aravindavk commented:
Let me know if any changes required.

quickfur commented on std/string.d:
```diff
@@ -3144,6 +3252,89 @@ if (isConvertibleToString!Range)
+    {
+    wstring s = " xyz ";
+    assert(s.sameTail(s.stripLeft(" ")));
+    assert(s.sameHead(s.stripRight(" ")));
```
Nitpick: please indent the body of the function literal here.

quickfur commented on std/string.d:
```diff
@@ -3144,6 +3252,89 @@ if (isConvertibleToString!Range)
+
+    assertCTFEable!(
+    {
+    static foreach (S; AliasSeq!( char[], const  char[],  string,
```
Come to think of it, this should also be indented too.

quickfur commented:
Actually, nevermind that, I see that you're just following the style in other parts of the same file. Can't say I'm a fan of this style, but since it's consistent, let's just roll with it.

JackStouffer commented:
Needs a changelog entry

aravindavk commented:
> Needs a changelog entry

added.

quickfur commented:
LGTM. Thanks!

quickfur commented:
ping @JackStouffer @wilzbach @burner @jmdavis 

Any more comments before I pull the trigger? ;-)

jmdavis commented:
Technically, I think that it requires @andralex's approval, because it's adding new public symbols.

quickfur commented:
ping @andralex

andralex commented:
Hm, I fear people might get confused. Consider: `strip("www.dlang.org", "www.")` whereby the user expects the `www.` to be stripped if present.

andralex commented on changelog/std-string-strip.dd:
```diff
@@ -0,0 +1,14 @@
+`std.string.strip` now accepts a list of characters to be stripped
```
"list" is not the right word

andralex commented on changelog/std-string-strip.dd:
```diff
@@ -0,0 +1,14 @@
+
+New overload functions of $(REF strip, std, string),
+$(REF stripLeft, std, string) and
+$(REF stripRight, std, string) accepts a list of characters to be
```
no list

andralex commented on std/string.d:
```diff
@@ -2887,14 +2890,64 @@ if (isConvertibleToString!Range)
+/// Ditto
+auto stripLeft(Range)(Range input, string chars)
+if (isForwardRange!Range && isSomeChar!(ElementEncodingType!Range) &&
+    !isInfinite!Range && !isConvertibleToString!Range)
```
range can be infinite

andralex commented on std/string.d:
```diff
@@ -3047,16 +3100,71 @@ if (isConvertibleToString!Range)
+/// Ditto
+auto stripRight(Range)(Range str, string chars)
+if (isBidirectionalRange!Range && isSomeChar!(ElementEncodingType!Range) &&
+    !isInfinite!Range && !isConvertibleToString!Range)
```
bidirectional and infinite not very likely

quickfur replied:
Hold on. How do you strip characters to the *right* of an infinite range?!

E.g., if the infinite range consists of `aaabaaabaaa...` repeated forever, and you call `stripRight` to remove `a`s from the right, how would the code know when to start dropping `a`s? And if `stripRight` was asked to remove *both* `a`s and `b`s, how would the code know if there's a `z` 500 elements into the range so that it shouldn't just return an empty range?

andralex replied:
The range has `back` and `popBack` and that seems to be all that's needed. It's difficult to define a range that is bidirectional and infinite.

quickfur replied:
Ahh, I misunderstood that you wanted to remove *both* bidirectional and infinite from the constraints.  Removing isInfinite and leaving bidirectional makes sense. :-)

quickfur replied:
P.S. It's not *that* hard to make an infinite range that's also bidirectional; the hard part is coming up with a sensible use case for it. :-P  The traditional infinite range is open to the right, i.e., it has no end. But it's also possible to have a range that's open in the middle, i.e., the front and back parts never meet. Such a range actually *would* work with this code, as is, if we drop `isInfinite`.  Though of course, I've a hard time imagining a real-world use case for this kind of bizarre infinite range!

andralex commented on std/string.d:
```diff
@@ -3047,16 +3100,71 @@ if (isConvertibleToString!Range)
+        if (indexOf(chars, c) == -1)
+            break;
+        str.popBack();
+    }
```
There's too much pomp and circumstance in this code. It's a simple for loop with no temporaries:

```d
for (; !str.empty; str.popBack)
{
    if (chars.indexOf(str.back) == -1)
        break;
}
```

andralex commented on std/string.d:
```diff
@@ -3144,6 +3252,89 @@ if (isConvertibleToString!Range)
+/// Ditto
+auto strip(Range)(Range str, string chars)
+if (isBidirectionalRange!Range && isSomeChar!(ElementEncodingType!Range) &&
+    !isInfinite!Range && !isConvertibleToString!Range)
```
drop isInfinite

quickfur replied:
Again, how do you strip characters from the right of an infinite range?

andralex replied:
With the code as is :)

quickfur replied:
Gotcha.

aravindavk commented:
Thanks @andralex for the review, reworked on the patch as suggested.

Also changed the second argument from string to `const(Char)[]`(similar to other functions in the same file) and added unit tests for the same.

aravindavk commented:
> Hm, I fear people might get confused. Consider: strip("www.dlang.org", "www.") whereby the user expects the www. to be stripped if present.

Many languages provides the same API(Python, PHP, Golang). 

In Go, `TrimPrefix` and `TrimSuffix` is added along with `TrimLeft` and `TrimRight` to avoid this confusion. Equivalant in D is `std.string.chompPrefix` (I couldn't find `std.string.chompSuffix`)

aravindavk commented:
jenkins/pr-merge test failed with following error. Is there a way I can retrigger this test?

    [diet-ng] Running shell script
    + dub build -b release
    Performing "release" build using /var/lib/jenkins/dlang_projects/distribution/bin/dmd for x86_64.
    diet-ng 1.4.4: building configuration "library"...
    Error: Error writing file '.dub/build/library-release-linux.posix-x86_64-dmd_2078-47FD7784A9F78ED910EB095685955196/libdiet-ng.a'
    /var/lib/jenkins/dlang_projects/distribution/bin/dmd failed with exit code 1.
    script returned exit code 2

jmdavis commented:
> Is there a way I can retrigger this test?

It triggers on new commits. So, just do a `git commit --amend` with no changes to the message, and then `git push -f` to your github repo, and it'll rerun everything.

wilzbach commented:
> Is there a way I can retrigger this test?

Only people with write permissions can login at Jenkins -> I just restarted Jenkins for you.
You could do it yourself by changing your commit sha:

```bash
git commit --amend
git push -f
```

Note that we currently have a couple of issues with Jenkins. The one you saw is https://github.com/dlang/ci/issues/135
tl;dr: we are aware of the Jenkins failures and they won't block your PR.

aravindavk commented:
Thanks a lot @wilzbach @jmdavis

quickfur commented:
LGTM

aravindavk commented:
Thanks for all the review comments. Let me know if any change required.

@andralex please review the changes. (jenkins/pr-merge failure is not related to the patch https://github.com/dlang/phobos/pull/6023#issuecomment-360804197)

andralex commented on std/string.d:
```diff
@@ -2887,14 +2890,69 @@ if (isConvertibleToString!Range)
+    assert(stripLeft(" hello ", "") == " hello ");
+}
+
+auto stripLeft(Range, Char)(auto ref Range str, const(Char)[] chars)
```
Public and undocumented not good. Please merge constraints and bodies with the overload above.

aravindavk replied:
I followed similar to existing `stripLeft(Range)(auto ref Range str)`. Confused because none of the functions with `auto ref` input are documented in this file. Please clarify.

andralex replied:
We pursue lifting the quality of code overall, and in the process of doing that it's possible that new code is subject to a higher bar than existing code. Two standards arose from experience:

* All public symbols must be documented
* Prefer fewer overloads

I understand this can be frustrating - please bear with us. Thanks!

aravindavk replied:
No problem, happy to work on the changes, just needed a clarification. Thanks

andralex commented on std/string.d:
```diff
@@ -3047,16 +3105,72 @@ if (isConvertibleToString!Range)
+           "     hello world");
+}
+
+auto stripRight(Range, Char)(auto ref Range str, const(Char)[] chars)
```
please merge with the other overload

andralex commented on std/string.d:
```diff
@@ -3144,6 +3258,92 @@ if (isConvertibleToString!Range)
+    assert(strip(" hello ", "") == " hello ");
+}
+
+auto strip(Range, Char)(auto ref Range str, const(Char)[] chars)
```
please merge with the other oveload

aravindavk commented:
@andralex merged the overloads as suggested. Please review.

aravindavk commented:
ci/circleci is failing due to following error.

    posix.mak:599: recipe for target 'std/traits.publictests' failed
    make: *** [std/traits.publictests] Hangup
    make: *** Deleting intermediate file 'generated/linux/debug/64/publictests/.directory'
    Too long with no output (exceeded 10m0s)

Is this a known failure or do I need to do anything to fix this?

wilzbach commented:
Known

wilzbach commented:
> Known

And fixed: https://github.com/dlang/phobos/pull/6099

wilzbach commented on std/string.d:
```diff
@@ -3080,16 +3137,71 @@ if (isConvertibleToString!Range)
+    {
+        for (; !str.empty; str.popBack)
+        {
+            if (chars.indexOf(str.back) == -1)
```
Why don't you use `foreach` here?
In this case there's _any_ performance penalty of this abstraction.
https://github.com/dlang/phobos/pull/6092#issuecomment-361397749

quickfur replied:
ditto

quickfur replied:
Here we're iterating from the back, so we'd need `foreach_reverse` or `retro` if you wanted to use an abstraction.

Or we could write this as:
````
while (!str.empty && chars.indexOf(str.back) != -1)
    str.popBack;
````
which to me reads more intuitively as "while the range is not empty and the back is a character to the stripped, pop the back".

aravindavk replied:
Shall I change this to use `while` as suggested?

andralex replied:
Not important, but I like the `for` form. Kurz und gut.

wilzbach replied:
> Not important, but I like the for form. Kurz und gut.

Yes, I was referring to `foreach_reverse`:

```d
foreach_reverse (s; str)
    if (chars.indexOf(s) == -1)
        break;
```

FWIW LDC optimizes `foreach_reverse` a lot better:
 
https://godbolt.org/g/k9tr4S

```asm
_D7example3fooFAyaQdZv:
  sub rsp, 72
  mov qword ptr [rsp + 64], rcx
  mov qword ptr [rsp + 56], rdx
  mov qword ptr [rsp + 40], rdi
  mov qword ptr [rsp + 48], rsi
  mov rcx, qword ptr [rsp + 64]
  mov rdx, qword ptr [rsp + 56]
  mov qword ptr [rsp + 24], rdx
  mov qword ptr [rsp + 32], rcx
  mov rcx, qword ptr [rsp + 24]
  mov qword ptr [rsp + 16], rcx
.LBB0_1:
  mov rax, qword ptr [rsp + 16]
  mov rcx, rax
  sub rcx, 1
  mov qword ptr [rsp + 16], rcx
  cmp rax, 0
  je .LBB0_6
  mov eax, 1
  mov rcx, qword ptr [rsp + 16]
  mov rdx, qword ptr [rsp + 32]
  mov sil, byte ptr [rdx + rcx]
  mov byte ptr [rsp + 15], sil
  movzx esi, byte ptr [rsp + 15]
  mov rdx, qword ptr [rsp + 40]
  mov rcx, qword ptr [rsp + 48]
  mov edi, 1
  mov dword ptr [rsp + 8], eax
  call _D3std6string__T7indexOfTAyaZQnFNaNbNiNfQpxwxEQBs8typecons__T4FlagVQBqa13_6361736553656e736974697665ZQBoZl@PLT
  cmp rax, -1
  jne .LBB0_4
  jmp .LBB0_6
.LBB0_4:
  jmp .LBB0_5
.LBB0_5:
  jmp .LBB0_1
.LBB0_6:
  add rsp, 72
  ret
```

I hope you notice that it could inline `back` completely.

For comparison, the `for`:

```asm
_D7example3fooFAyaQdZv:
  push r15
  push r14
  push r12
  push rbx
  sub rsp, 24
  mov r14, rsi
  mov r12, rdi
  mov qword ptr [rsp + 8], rdx
  mov qword ptr [rsp + 16], rcx
  test rdx, rdx
  je .LBB0_9
  mov rdi, rdx
  mov rsi, rcx
  call _D3std5range10primitives__T4backTyaZQjFNaNdNfAyaZw@PLT
  mov edi, 1
  mov esi, eax
  mov rdx, r12
  mov rcx, r14
  call _D3std6string__T7indexOfTAyaZQnFNaNbNiNfQpxwxEQBs8typecons__T4FlagVQBqa13_6361736553656e736974697665ZQBoZl@PLT
  cmp rax, -1
  je .LBB0_9
  lea r15, [rsp + 8]
.LBB0_3:
  mov rbx, qword ptr [rsp + 8]
  test rbx, rbx
  je .LBB0_6
  mov rdi, rbx
  mov rsi, r15
  call _D3std3utf__T10strideBackTAyaZQrFNaNfKQmmZk@PLT
  mov eax, eax
  sub rbx, rax
  cmp rbx, qword ptr [rsp + 8]
  ja .LBB0_5
  mov qword ptr [rsp + 8], rbx
  test rbx, rbx
  je .LBB0_9
  mov rsi, qword ptr [rsp + 16]
  mov rdi, rbx
  call _D3std5range10primitives__T4backTyaZQjFNaNdNfAyaZw@PLT
  mov edi, 1
  mov esi, eax
  mov rdx, r12
  mov rcx, r14
  call _D3std6string__T7indexOfTAyaZQnFNaNbNiNfQpxwxEQBs8typecons__T4FlagVQBqa13_6361736553656e736974697665ZQBoZl@PLT
  cmp rax, -1
  jne .LBB0_3
.LBB0_9:
  add rsp, 24
  pop rbx
  pop r12
  pop r14
  pop r15
  ret
.LBB0_6:
  lea rsi, [rip + .L.str]
  lea rcx, [rip + .L.str.1]
  mov edi, 69
  mov edx, 92
  mov r8d, 2206
  call _d_assert_msg@PLT
.LBB0_5:
  lea rsi, [rip + .L.str.1]
  mov edi, 92
  mov edx, 2207
  call _d_arraybounds@PLT
```

 If we would have perfect abstraction, I would have suggested `str.retro.until!(s => chars.indexOf(s) > 0);`, but unfortunately that's - even for the LLVM optimizer - too much.

wilzbach commented on std/string.d:
```diff
@@ -2920,14 +2923,68 @@ if (isConvertibleToString!Range)
+        return stripLeft!(StringTypeOf!Range)(input, chars);
+    else
+    {
+        for (; !input.empty; input.popFront)
```
Why don't you use `foreach` here?
In this case there's _any_ performance penalty of this abstraction.
https://github.com/dlang/phobos/pull/6092#issuecomment-361397749

quickfur replied:
I think the catch here is that we need to iterate *from the back*, not from the front. So we'll either have to use `foreach_reverse` (*ick*), or `std.range.retro`, which then would require passing the range in by reference, since we need to guarantee that the range is actually modified in order to return it to the user. So it'll end up being something ugly like this:
````
import std.range : retro;
foreach (ch; (&input).retro) { ... }
````
which is equally *ick*.

quickfur replied:
Oh wait, what I said only applies to the `stripRight` variant.  Here we're moving forwards, so there's no `foreach_reverse` or `retro` needed.

I'd say the `for` loop here is justified (using foreach would require passing the range by ref, just in case, so it still entails the ugly `&range` syntax).  Perhaps a more self-documenting way to write this would be:
````
while (!input.empty && chars.indexOf(input.front) != -1)
    input.popFront;
````
which, at least to me, reads more intuitively as "while the range is not empty and the front element is a character to be stripped, pop the front element".

wilzbach replied:
Well, it boils down to whether we want to do auto-decoding or not ;-)

aravindavk commented:
> And fixed: #6099 (I restarted your build on CircleCi)

Thanks @quickfur

andralex commented on std/string.d:
```diff
@@ -3177,6 +3289,89 @@ if (isConvertibleToString!Range)
 }
 
+/// Ditto
+auto strip(Range, Char)(Range str, const(Char)[] chars)
```
An additional overload would take a third argument:

```d
auto strip(Range, Char)(Range str, const(Char)[] leftChars, const(Char)[] rightChars)
```

and would strip `leftChars` on the left and `rightChars` on the right.

aravindavk replied:
Nice. I will add it

aravindavk replied:
Added the new overload as suggested and added the respective tests.

wilzbach commented on std/string.d:
```diff
@@ -2920,14 +2923,68 @@ if (isConvertibleToString!Range)
+///
+@safe pure unittest
+{
+     assert(stripLeft("     hello world     ", " ") ==
```
Indentation

aravindavk replied:
fixed

wilzbach commented on std/string.d:
```diff
@@ -2920,14 +2923,68 @@ if (isConvertibleToString!Range)
+    assert(stripLeft("xxxyy    hello world     ", "xy ") ==
+           "hello world     ");
+
+    import std.array : array;
```
This is a good sign that it should be a separate unittest.

aravindavk replied:
done

aravindavk commented:
@andralex @quickfur @wilzbach is this patch looking good now. 

Thanks

quickfur commented:
Alright, I think this is good to go.  Any further fixes can be done in followup PRs. Merge ahoy!

aravindavk commented:
Feeling very happy to see my first contribution to D language :)

Contribution experience was very nice, review comments were very valid and helped me to understand the language better. Happy to contribute more to D language in future. Thanks a lot

ZombineDev commented:
Nice work @aravindavk, you made a solid contribution to the standard library!

wilzbach commented:
> Feeling very happy to see my first contribution to D language :)

BTW it's online on the changelog already:

https://dlang.org/changelog/pending.html#std-string-strip

> Happy to contribute more to D language in future.

:+1: 
I have sent out an invite to our Slack channel too (it's a good way to quickly ask questions) - you deserve d it!

Also, https://issues.dlang.org is open 24/7 :wink:
